From 67b2e6435568649f352c08350e09b7d50274a00d Mon Sep 17 00:00:00 2001
From: "Darryl L. Pierce" <mcpierce@apache.org>
Date: Fri, 10 May 2013 14:20:35 +0000
Subject: [PATCH 2/6] QPID-4826: Patch Perl bindings memory leak

Marked the VariantToPerl() method in swig_perl_typemaps.i as returning a
new object, which indicates to the Perl bindings that it needs to keep
track of the return value for garbage collection.

Also removed the Swig 1.3.32 minimum restriction to enable building on
RHEL5 systems.

git-svn-id: https://svn.apache.org/repos/asf/qpid/trunk@1481021 13f79535-47bb-0310-9956-ffa450edef68
---
 qpid/cpp/bindings/CMakeLists.txt           | 50 +++++++++++++-----------------
 qpid/cpp/include/qpid/swig_perl_typemaps.i |  1 +
 2 files changed, 23 insertions(+), 28 deletions(-)

diff --git a/qpid/cpp/bindings/CMakeLists.txt b/qpid/cpp/bindings/CMakeLists.txt
index 2d2f74c..bab4db0 100644
--- a/qpid/cpp/bindings/CMakeLists.txt
+++ b/qpid/cpp/bindings/CMakeLists.txt
@@ -69,37 +69,32 @@ if ((${CMAKE_MAJOR_VERSION} EQUAL 2) AND (${CMAKE_MINOR_VERSION} LESS 8))
   endif (PERL_FOUND)
 endif ((${CMAKE_MAJOR_VERSION} EQUAL 2) AND (${CMAKE_MINOR_VERSION} LESS 8))
 
-set (SWIG_MINIMUM_VERSION "1.3.32")
-
 if (SWIG_FOUND)
-  if (${SWIG_VERSION} VERSION_LESS ${SWIG_MINIMUM_VERSION})
-    message("Found Swig < ${SWIG_MINIMUM_VERSION} - skipping language bindings")
-  else()
-    set(CMAKE_SWIG_FLAGS "-w361,362,401,467,503")
+  set(CMAKE_SWIG_FLAGS "-w361,362,401,467,503")
 
-    if (PYTHONLIBS_FOUND)
-        message("Building Python bindings")
-        execute_process(COMMAND ${PYTHON_EXECUTABLE}
-                        -c "from distutils.sysconfig import get_python_lib; print get_python_lib(True, prefix='${CMAKE_INSTALL_PREFIX}')"
-                        OUTPUT_VARIABLE PYTHON_SITEARCH_PACKAGES
-                        OUTPUT_STRIP_TRAILING_WHITESPACE)
+  if (PYTHONLIBS_FOUND)
+      message("Building Python bindings")
+      execute_process(COMMAND ${PYTHON_EXECUTABLE}
+                      -c "from distutils.sysconfig import get_python_lib; print get_python_lib(True, prefix='${CMAKE_INSTALL_PREFIX}')"
+                      OUTPUT_VARIABLE PYTHON_SITEARCH_PACKAGES
+                      OUTPUT_STRIP_TRAILING_WHITESPACE)
 
-        add_subdirectory(qpid/python)
-        add_subdirectory(qmf2/python)
-        add_subdirectory(qmf/python)
-    endif (PYTHONLIBS_FOUND)
+      add_subdirectory(qpid/python)
+      add_subdirectory(qmf2/python)
+      add_subdirectory(qmf/python)
+  endif (PYTHONLIBS_FOUND)
 
-    if (RUBY_FOUND)
-        message("Building Ruby bindings")
-        execute_process(COMMAND ${RUBY_EXECUTABLE} -r rbconfig -e "puts RbConfig::CONFIG['prefix']"
-                        OUTPUT_VARIABLE RUBY_PREFIX
-                        OUTPUT_STRIP_TRAILING_WHITESPACE)
-        string(REPLACE ${RUBY_PREFIX} ${CMAKE_INSTALL_PREFIX} RUBY_PFX_ARCH_DIR ${RUBY_SITEARCH_DIR})
-#       string(REPLACE ${RUBY_PREFIX} ${CMAKE_INSTALL_PREFIX} RUBY_PFX_ARCH_DIR ${RUBY_ARCH_DIR})
-        add_subdirectory(qpid/ruby)
-        add_subdirectory(qmf2/ruby)
-        add_subdirectory(qmf/ruby)
-    endif (RUBY_FOUND)
+  if (RUBY_FOUND)
+      message("Building Ruby bindings")
+      execute_process(COMMAND ${RUBY_EXECUTABLE} -r rbconfig -e "puts RbConfig::CONFIG['prefix']"
+                      OUTPUT_VARIABLE RUBY_PREFIX
+                      OUTPUT_STRIP_TRAILING_WHITESPACE)
+      string(REPLACE ${RUBY_PREFIX} ${CMAKE_INSTALL_PREFIX} RUBY_PFX_ARCH_DIR ${RUBY_SITEARCH_DIR})
+#     string(REPLACE ${RUBY_PREFIX} ${CMAKE_INSTALL_PREFIX} RUBY_PFX_ARCH_DIR ${RUBY_ARCH_DIR})
+      add_subdirectory(qpid/ruby)
+      add_subdirectory(qmf2/ruby)
+      add_subdirectory(qmf/ruby)
+  endif (RUBY_FOUND)
 
   if (PERLLIBS_FOUND)
       message("Building Perl bindings")
@@ -111,5 +106,4 @@ if (SWIG_FOUND)
 
       add_subdirectory(qpid/perl)
   endif (PERLLIBS_FOUND)
-  endif (${SWIG_VERSION} VERSION_LESS ${SWIG_MINIMUM_VERSION})
 endif (SWIG_FOUND)
diff --git a/qpid/cpp/include/qpid/swig_perl_typemaps.i b/qpid/cpp/include/qpid/swig_perl_typemaps.i
index c1e1d53..da24bfe 100644
--- a/qpid/cpp/include/qpid/swig_perl_typemaps.i
+++ b/qpid/cpp/include/qpid/swig_perl_typemaps.i
@@ -17,6 +17,7 @@
  * under the License.
  */
 
+%newobject VariantToPerl;
 %wrapper %{
 
 #include <stdarg.h>
-- 
1.8.1.4

